Merge upstream 5.6.36 into Percona Server

percona-ysorokin opened the pull request.

percona-ysorokin commented:
Jenkins build link
http://jenkins.percona.com/view/PS%205.6/job/percona-server-5.6-param/1852/

percona-ysorokin commented:
@laurynas-biveinis Please double-check the following:

1. Changes related to changing `os_file_t` to `pfs_os_file_t`. In particular, changing the type of the `file` field in `log_online_bitmap_file_struct`.
2. Changes related to `USE_NEW_XLINE_INTERFACE`.
3. Combined `const bool	is_last` condition in `logs_empty_and_mark_files_at_shutdown()` in `log0log.cc`

laurynas-biveinis commented on storage/innobase/log/log0log.cc:
```diff
@@ -3675,7 +3681,7 @@ logs_empty_and_mark_files_at_shutdown(void)
 
 	ut_a(lsn == log_sys->lsn);
-	ut_ad(lsn == log_sys->last_checkpoint_lsn);
+	ut_ad(srv_force_recovery == SRV_FORCE_NO_LOG_REDO || lsn == log_sys->last_checkpoint_lsn);
```
I'd slightly prefer >= SRV_FORCE_NO_LOG_REDO in the case a higher recovery level is added

laurynas-biveinis commented on storage/innobase/log/log0online.cc:
```diff
@@ -1768,8 +1768,9 @@ log_online_bitmap_iterator_next(
 			i->in_i++;
-			success = os_file_close_no_error_handling(i->in.file);
-			i->in.file = os_file_invalid;
+			success = os_file_close_no_error_handling(
```
os_file_close_no_error_handling needs PFS treatment parallel to that of os_file_close, and here file should be passed, not file.m_file

laurynas-biveinis commented on storage/innobase/log/log0online.cc:
```diff
@@ -564,9 +564,9 @@ log_online_rotate_bitmap_file(
 					part */
 {
-	if (log_bmp_sys->out.file != os_file_invalid) {
+	if (log_bmp_sys->out.file.m_file != os_file_invalid) {
```
Please remove os_file_invalid altogether (it was introduced by us for log0online), and make a helper function

laurynas-biveinis commented on storage/innobase/log/log0online.cc:
```diff
@@ -1172,7 +1172,7 @@ log_online_write_bitmap_page(
 
 #ifdef UNIV_LINUX
-	posix_fadvise(log_bmp_sys->out.file, log_bmp_sys->out.offset,
+	posix_fadvise(log_bmp_sys->out.file.m_file, log_bmp_sys->out.offset,
```
The abstraction level is off, please create a pfs_os_file-taking function in os0file.h. Probably OK to leave abstraction level violation in keeping POSIX_FADV_ macros at the callers though. #ifdef UNIV_LINUX should go inside the new function implementation

laurynas-biveinis commented on storage/innobase/log/log0online.cc:
```diff
@@ -1878,7 +1879,7 @@ log_online_purge_changed_page_bitmaps(
+		log_bmp_sys->out.file.m_file = os_file_invalid;
 	}
 
 	for (i = 0; i < bitmap_files.count; i++) {
```
In the end, log0online.cc|h|i should be clean of direct m_file references

percona-ysorokin commented:
Added the following macros / pfs functions to the InnoDB OS file layer:
'os_file_close_no_error_handling_pfs()'(instrumented under 'PSI_FILE_CLOSE');
'os_file_advise_pfs()' (instrumented under 'PSI_FILE_SEEK');
'os_file_set_eof_at_pfs()' (instrumented under 'PSI_FILE_CHSIZE');
'os_file_allocate_pfs()' (instrumented under 'PSI_FILE_CHSIZE').
In order to keep ABI compatibility, 'enum PSI_file_operation' was not
extended with new values. Therefore, the closest by meaning existing ones
were used for the new instrumented functions.

Added 'os_file_is_invalid()' and 'os_file_mark_invalid()' which give high
level abstractions for invalid file handles.

'log0online.cc' reworked using new PSI functions / macros.

Updated Jenkins build link
http://jenkins.percona.com/view/PS%205.6/job/percona-server-5.6-param/1859/

Jenkins build with `-DWITH_PERFSCHEMA_STORAGE_ENGINE=OFF` cmake option
http://jenkins.percona.com/view/PS%205.6/job/percona-server-5.6-param/1860/
A number of tests fail there as they do not have `--source include/have_perfschema.inc`
(this issue will be submitted as a LP bug).

laurynas-biveinis commented:
I don't think 'os_file_advise_pfs()' (instrumented under 'PSI_FILE_SEEK'); is a correct thing to do - too unrelated. And advise probably does not need PFS annotation at all?

laurynas-biveinis commented on storage/innobase/include/os0file.h:
```diff
@@ -628,6 +788,56 @@ ibool
+	os_file_t	file,	/*!< in, own: handle to a file */
+	os_offset_t	offset,	/*!< in: file region offset  */
+	os_offset_t	len,	/*!< in: file region length  */
+	ulint		advice);/*!< in: advice for access pattern */
```
os_file_advise_t ?

percona-ysorokin replied:
Although, I did not find anything about this in the man page
http://man7.org/linux/man-pages/man2/posix_fadvise.2.html
In our own code(`row0merge.cc`) we have 
```c++
#ifdef POSIX_FADV_SEQUENTIAL
	/* The input file will be read sequentially, starting from the
	beginning and the middle.  In Linux, the POSIX_FADV_SEQUENTIAL
	affects the entire file.  Each block will be read exactly once. */
	posix_fadvise(file->fd, 0, 0,
		      POSIX_FADV_SEQUENTIAL | POSIX_FADV_NOREUSE);
#endif /* POSIX_FADV_SEQUENTIAL */
```
where we pass several ORed POSIX_FADV_XXX values.
So, just in case, I am passing `ulint` instead of `os_file_advise_t`

laurynas-biveinis replied:
Then, strictly speaking, the arg should be int, not ulint?

Up to you to decide. The only comment that has to be addressed is the one about fadvise not needing PFS

percona-ysorokin replied:
I just wanted to make it similar to `os_file_create` where `create_mode` parameter which can be a combination of `os_file_create_t` values is defined as `ulint`.

laurynas-biveinis commented on storage/innobase/include/os0file.h:
```diff
@@ -865,24 +1088,78 @@ pfs_os_file_delete_if_exists_func(
+	pfs_os_file_t	file,	/*!< in, own: handle to a file */
+	os_offset_t	offset,	/*!< in: file region offset  */
+	os_offset_t	len,	/*!< in: file region length  */
+	ulint		advice,	/*!< in: advice for access pattern */
```
os_file_advise_t ?

laurynas-biveinis commented on storage/innobase/include/os0file.ic:
```diff
@@ -449,4 +557,90 @@ pfs_os_file_delete_if_exists_func(
+	pfs_os_file_t	file,	/*!< in, own: handle to a file */
+	os_offset_t	offset,	/*!< in: file region offset  */
+	os_offset_t	len,	/*!< in: file region length  */
+	ulint		advice,	/*!< in: advice for access pattern */
```
os_file_advise_t ?

laurynas-biveinis commented on storage/innobase/os/os0file.cc:
```diff
@@ -2140,39 +2144,112 @@ os_file_close_no_error_handling(
+	os_file_t	file,	/*!< in, own: handle to a file */
+	os_offset_t	offset,	/*!< in: file region offset  */
+	os_offset_t	len,	/*!< in: file region length  */
+	ulint		advice)/*!< in: advice for access pattern */
```
os_file_advise_t ?

percona-ysorokin commented:
Reworked `os_file_advise()` - it's a regular function now.
Jenkins build link (with PFS)
http://jenkins.percona.com/job/percona-server-5.6-param/1861/
Jenkins build link (without PFS)
http://jenkins.percona.com/job/percona-server-5.6-param/1862/
